### PR TITLE
fix in cluster operator create namespace for istio control plane.

### LINKED ISSUE #29519: In-cluster operator should be able to create the namespace for istio control plane if it doesn't exist
**Bug description**

Currently if we initialize the istio operator with `--watchedNamespaces` to specify the istio operator watching namespace other than `istio-system`, and create the `istiooperator` CR in the watched namespace will results error `namespaces "istio-system" not found`

For example:
```
# istioctl operator init --watchedNamespaces=ns1,ns2
```

Then create istiooperator CR in ns1:
```
# cat iop.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: ns1
  name: example-istiocontrolplane
spec:
  profile: default
  namespace: istio-system
```

The In-cluster operator will report `namespaces "istio-system" not found` error.

In-cluster operator should be able to create the namespace for istio control plane if the `istiooperator` CR's namespace is different from the namespace of istio control plane.

This also align with the `istioctl`.

[ ] Docs
[ X ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**

**Steps to reproduce the bug**

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

**How was Istio installed?**

**Environment where the bug was observed (cloud vendor, OS, etc)**

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

### PR DESCRIPTION
**Pull request by morvencao**
resolve: https://github.com/istio/istio/issues/29519

[ ] Configuration Infrastructure
[ ] Docs
[ X ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: operator/pkg/helmreconciler/reconciler.go [hunk 1]
```diff
@@ -489,3 +495,53 @@ func (h *HelmReconciler) reportPrunedObjectKind() {
 			Increment()
 	}
 }
+
+// createNamespace creates a namespace using the given k8s client.
+func (h *HelmReconciler) createNamespace(namespace string, network string) error {
```
Comment (shamsher31) [inline operator/pkg/helmreconciler/reconciler.go:500]: Is it safe to remove `createNamespace` from `shared.go` as it is moved in `reconciler`? 
https://github.com/istio/istio/blob/b0474ef00481d31157bb9f4cab3f8690ef6fb6a9/operator/cmd/mesh/shared.go#L246

**Comment by shamsher31** [comment]
@morvencao Is this the right config to test with?

```
# iop.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: ns1
  name: example-istiocontrolplane
spec:
  profile: default
  namespace: ns1
```
```
$ ./out/linux_amd64/istioctl install -f iop.yaml
```

File: operator/pkg/helmreconciler/reconciler.go (hunk 1 above)
Comment (morvencao) [reply operator/pkg/helmreconciler/reconciler.go:500]: Looks like the `istioctl operator init -f <cr.yaml>` still needs that function.

**Comment by morvencao** [comment]
@shamsher31 yes, we also need to test the in-cluster operator.

**Comment by shamsher31** [comment]
```
$ ./out/linux_amd64/istioctl operator init --watchedNamespaces=ns1
Installing operator controller in namespace: istio-operator using image: gcr.io/istio-testing/operator:1.9-dev
Operator controller will watch namespaces: ns1
✔ Istio operator installed                                                                                                              
✔ Installation complete
```
```
# iop.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: ns1
  name: example-istiocontrolplane
spec:
  profile: default
  namespace: ns1
```

It got stuck with the following error.

```
$ ./out/linux_amd64/istioctl install -f iop.yaml
This will install the Istio 1.9.0 default profile with ["Istio core" "Istiod" "Ingress gateways"] components into the cluster. Proceed? (y/N) y
2020-12-11T10:17:15.706573Z     error   installer       failed to update resource with server-side apply for obj ServiceAccount/istio-system/istio-reader-service-account: namespaces "istio-system" not found
2020-12-11T10:17:15.760166Z     error   installer       failed to update resource with server-side apply for obj ServiceAccount/istio-system/istiod-service-account: namespaces "istio-system" not found
- Processing resources for Istio core.                                                                                                  2020-12-11T10:17:15.840996Z     error   installer       failed to update resource with server-side apply for obj Role/istio-system/istiod-istio-system: namespaces "istio-system" not found
2020-12-11T10:17:15.905859Z     error   installer       failed to update resource with server-side apply for obj RoleBinding/istio-system/istiod-istio-system: namespaces "istio-system" not found
✘ Istio core encountered an error: failed to update resource with server-side apply for obj ServiceAccount/istio-system/istio-reader-service-account: namespaces "istio-system" not found
failed to update resource with server-side apply for obj ServiceAccount/istio-system/istiod-service-account: namespaces "istio-system" not found
failed to update resource with server-side apply for obj Role/istio-system/istiod-istio-system: namespaces "istio-system" not found
failed to update resource with server-side apply for obj RoleBinding/istio-system/istiod-istio-system: namespaces "istio-system" not found
2020-12-11T10:17:15.996240Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/metadata-exchange-1.8: namespaces "istio-system" not found
2020-12-11T10:17:16.050809Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/metadata-exchange-1.9: namespaces "istio-system" not found
2020-12-11T10:17:16.105218Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/stats-filter-1.8: namespaces "istio-system" not found
2020-12-11T10:17:16.159676Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/stats-filter-1.9: namespaces "istio-system" not found
2020-12-11T10:17:16.214687Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-metadata-exchange-1.8: namespaces "istio-system" not found
2020-12-11T10:17:16.268885Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-metadata-exchange-1.9: namespaces "istio-system" not found
2020-12-11T10:17:16.323775Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-stats-filter-1.8: namespaces "istio-system" not found
2020-12-11T10:17:16.378471Z     error   installer       failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-stats-filter-1.9: namespaces "istio-system" not found
✘ Istiod encountered an error: failed to update resource with server-side apply for obj EnvoyFilter/istio-system/metadata-exchange-1.8: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/metadata-exchange-1.9: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/stats-filter-1.8: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/stats-filter-1.9: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-metadata-exchange-1.8: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-metadata-exchange-1.9: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-stats-filter-1.8: namespaces "istio-system" not found
failed to update resource with server-side apply for obj EnvoyFilter/istio-system/tcp-stats-filter-1.9: namespaces "istio-system" not found
- Processing resources for Ingress gateways. Waiting for Deployment/ns1/istio-ingressgateway  
```

**Comment by morvencao** [comment]
@shamsher31 See: https://github.com/istio/istio/issues/26813 for more details:
Looks likes we still need:
```
# cat iop.yaml
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: ns1
  name: example-istiocontrolplane
spec:
  values:
    global:
      istioNamespace: ns1
    meshConfig:
      rootNamespace: ns1
```

**Comment by shamsher31** [comment]
@morvencao I confirmed and it worked after adding  `meshConfig.rootNamespace`.

```
$ ./out/linux_amd64/istioctl install -f iop.yaml
This will install the Istio 1.9.0 default profile with ["Istio core" "Istiod" "Ingress gateways"] components into the cluster. Proceed? (y/N) y
✔ Istio core installed                                                                                                              
✔ Istiod installed                                                                                                                  
✔ Ingress gateways installed                                                                                                        
✔ Installation complete
```

I still think there are a lot of `namespace` options we need to set in order for the operator to work properly. may be in future we should optimize it.

**Review by shamsher31 — APPROVED** [review]
Thanks for the fix.

**Review by irisdingbj — APPROVED** [review]

**Pull request merged**